lib/semantics/expression.h

```diff
@@ -27,30 +29,100 @@ template<typename> struct Integer;
 template<typename> struct Constant;
 }
 
-namespace Fortran::semantics {
+// The expression semantic analysis code has its implementation in
+// namespace Fortran::evaluate, but the exposed API to it is in the
+// namespace Fortran::semantics (below).
+//
+// The template function AnalyzeExpr is an internal interface
+// between the implementation and the API used by semantic analysis.
+// This template function has a few specializations here in the header
+// file to handle what semantics might want to pass in as a top-level
+// expression; other specializations appear in the implementation.
+//
+// The ExpressionAnalysisContext wraps a SemanticsContext reference
+// and implements constraint checking on expressions using the
+// parse tree node wrappers that mirror the grammar annotations used
+// in the Fortran standard (i.e., scalar-, constant-, &c.).  These
+// constraint checks are performed in a deferred manner so that any
+// errors are reported on the most accurate source location available.
 
-class SemanticsContext;
+namespace Fortran::evaluate {
+class ExpressionAnalysisContext {
+public:
+  using ConstraintChecker = bool (ExpressionAnalysisContext::*)(
+      Expr<SomeType> &);
 
-// Semantic analysis of one expression.
-std::optional<evaluate::Expr<evaluate::SomeType>> AnalyzeExpr(
-    SemanticsContext &, const parser::Expr &);
+  ExpressionAnalysisContext(semantics::SemanticsContext &sc) : context_{sc} {}
+  ExpressionAnalysisContext(ExpressionAnalysisContext &i)
+    : context_{i.context_}, inner_{&i} {}
+  ExpressionAnalysisContext(ExpressionAnalysisContext &i, ConstraintChecker cc)
+    : context_{i.context_}, inner_{&i}, constraint_{cc} {}
 
+  semantics::SemanticsContext &context() const { return context_; }
+
+  template<typename... A> void Say(A... args) {
+    context_.foldingContext().messages.Say(std::forward<A>(args)...);
+  }
+
+  void CheckConstraints(std::optional<Expr<SomeType>> &);
+  bool ScalarConstraint(Expr<SomeType> &);
+  bool ConstantConstraint(Expr<SomeType> &);
+  bool IntegerConstraint(Expr<SomeType> &);
```

Comment:
Need logical?

Reply:
Haven't yet (nor CHARACTER) but I'll deal with it now.